File: packages/prism-core/src/language/syntax/case-utils.ts
/**
 * Case-conversion utilities for codegen.
 *
 * Shared by all emitters and Prism packages. Single source of truth for
 * toCamel/toPascal/toScreamingSnake/toSnakeCase conversions.
 */

// ── Internal: split into words ──────────────────────────────────────────────

/**
 * Split a string into words, handling:
 *   snake_case, kebab-case, dot.case, camelCase, PascalCase, SCREAMING_SNAKE
 */
function splitWords(s: string): string[] {
  return s
    .replace(/[^a-zA-Z0-9]/g, ' ')              // snake_case, kebab-case, dot.case → spaces
    .replace(/([a-z])([A-Z])/g, '$1 $2')          // camelCase boundary
    .replace(/([A-Z]+)([A-Z][a-z])/g, '$1 $2')    // HTMLParser → HTML Parser
    .trim()
    .split(/\s+/)
    .filter(Boolean);
}

// ── Public API ──────────────────────────────────────────────────────────────

/** Strip non-identifier characters, ensure starts with letter or underscore. */
export function safeIdentifier(s: string): string {
  const clean = s.replace(/[^a-zA-Z0-9_]/g, '_').replace(/^([0-9])/, '_$1');
  return clean || '_unknown';
}

/** Convert any case to camelCase. */
export function toCamelCase(s: string): string {
  const words = splitWords(s);
  if (words.length === 0) return '_unknown';
  return words
    .map((w, i) => i === 0 ? w.toLowerCase() : w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
    .join('');
}

/** Convert any case to PascalCase. */
export function toPascalCase(s: string): string {
  const words = splitWords(s);
  if (words.length === 0) return '_Unknown';
  return words
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase())
    .join('');
}

/** Convert any case to SCREAMING_SNAKE_CASE. */
export function toScreamingSnake(s: string): string {
  const words = splitWords(s);
  if (words.length === 0) return '_UNKNOWN';
  return words.map((w) => w.toUpperCase()).join('_');
}

/** Convert any case to snake_case. */
export function toSnakeCase(s: string): string {
  const words = splitWords(s);
  if (words.length === 0) return '_unknown';
  return words.map((w) => w.toLowerCase()).join('_');
}

/** safeIdentifier + toCamelCase */
export function toCamelIdent(s: string): string {
  return toCamelCase(safeIdentifier(s));
}

/** safeIdentifier + toPascalCase */
export function toPascalIdent(s: string): string {
  return toPascalCase(safeIdentifier(s));
}

/** safeIdentifier + toScreamingSnake */
export function toScreamingSnakeIdent(s: string): string {
  return toScreamingSnake(safeIdentifier(s));
}

File: packages/prism-core/src/language/syntax/case_utils.rs
//! Case-conversion utilities for codegen.
//!
//! Port of `language/syntax/case-utils.ts`. Single source of truth
//! for `to_camel_case` / `to_pascal_case` / `to_screaming_snake` /
//! `to_snake_case` plus their safe-identifier variants.

/// Split a string into words, handling snake_case, kebab-case,
/// dot.case, camelCase, PascalCase, and SCREAMING_SNAKE.
fn split_words(s: &str) -> Vec<String> {
    // First collapse non-alphanumerics to spaces.
    let cleaned: String = s
        .chars()
        .map(|c| if c.is_ascii_alphanumeric() { c } else { ' ' })
        .collect();

    // Walk with camelCase-boundary + ACRONYMBoundary splitting.
    let chars: Vec<char> = cleaned.chars().collect();
    let mut pieces: Vec<String> = Vec::new();
    let mut current = String::new();

    for i in 0..chars.len() {
        let ch = chars[i];
        if ch == ' ' {
            if !current.is_empty() {
                pieces.push(std::mem::take(&mut current));
            }
            continue;
        }

        if !current.is_empty() {
            let prev = chars[i - 1];
            // camelCase boundary: aB
            let camel_boundary = prev.is_ascii_lowercase() && ch.is_ascii_uppercase();
            // Acronym boundary: HTMLParser → HTML|Parser.
            // Break when we go UPPER, UPPER, lower (split before the
            // second upper).
            let acronym_boundary = prev.is_ascii_uppercase()
                && ch.is_ascii_uppercase()
                && i + 1 < chars.len()
                && chars[i + 1].is_ascii_lowercase();
            if camel_boundary || acronym_boundary {
                pieces.push(std::mem::take(&mut current));
            }
        }
        current.push(ch);
    }
    if !current.is_empty() {
        pieces.push(current);
    }

    pieces.into_iter().filter(|w| !w.is_empty()).collect()
}

fn capitalize(word: &str) -> String {
    let mut chars = word.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => {
            let rest: String = chars.as_str().to_lowercase();
            first.to_ascii_uppercase().to_string() + &rest
        }
    }
}

/// Strip non-identifier characters; ensure the result starts with
/// a letter or underscore.
pub fn safe_identifier(s: &str) -> String {
    let mut out: String = s
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '_' {
                c
            } else {
                '_'
            }
        })
        .collect();
    if out.is_empty() {
        return "_unknown".to_string();
    }
    if out.chars().next().is_some_and(|c| c.is_ascii_digit()) {
        out.insert(0, '_');
    }
    out
}

pub fn to_camel_case(s: &str) -> String {
    let words = split_words(s);
    if words.is_empty() {
        return "_unknown".to_string();
    }
    words
        .into_iter()
        .enumerate()
        .map(|(i, w)| {
            if i == 0 {
                w.to_lowercase()
            } else {
                capitalize(&w)
            }
        })
        .collect()
}

pub fn to_pascal_case(s: &str) -> String {
    let words = split_words(s);
    if words.is_empty() {
        return "_Unknown".to_string();
    }
    words.iter().map(|w| capitalize(w)).collect()
}

pub fn to_screaming_snake(s: &str) -> String {
    let words = split_words(s);
    if words.is_empty() {
        return "_UNKNOWN".to_string();
    }
    words
        .into_iter()
        .map(|w| w.to_uppercase())
        .collect::<Vec<_>>()
        .join("_")
}

pub fn to_snake_case(s: &str) -> String {
    let words = split_words(s);
    if words.is_empty() {
        return "_unknown".to_string();
    }
    words
        .into_iter()
        .map(|w| w.to_lowercase())
        .collect::<Vec<_>>()
        .join("_")
}

pub fn to_camel_ident(s: &str) -> String {
    to_camel_case(&safe_identifier(s))
}

pub fn to_pascal_ident(s: &str) -> String {
    to_pascal_case(&safe_identifier(s))
}

pub fn to_screaming_snake_ident(s: &str) -> String {
    to_screaming_snake(&safe_identifier(s))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn camel_case_conversions() {
        assert_eq!(to_camel_case("hello_world"), "helloWorld");
        assert_eq!(to_camel_case("HelloWorld"), "helloWorld");
        assert_eq!(to_camel_case("hello-world"), "helloWorld");
        assert_eq!(to_camel_case("hello.world"), "helloWorld");
        assert_eq!(to_camel_case("HTMLParser"), "htmlParser");
    }

    #[test]
    fn pascal_case_conversions() {
        assert_eq!(to_pascal_case("hello_world"), "HelloWorld");
        assert_eq!(to_pascal_case("helloWorld"), "HelloWorld");
        assert_eq!(to_pascal_case("HTMLParser"), "HtmlParser");
    }

    #[test]
    fn screaming_snake_conversions() {
        assert_eq!(to_screaming_snake("helloWorld"), "HELLO_WORLD");
        assert_eq!(to_screaming_snake("hello-world"), "HELLO_WORLD");
    }

    #[test]
    fn snake_case_conversions() {
        assert_eq!(to_snake_case("helloWorld"), "hello_world");
        assert_eq!(to_snake_case("HelloWorld"), "hello_world");
        assert_eq!(to_snake_case("HTMLParser"), "html_parser");
    }

    #[test]
    fn safe_identifier_leading_digit() {
        assert_eq!(safe_identifier("123abc"), "_123abc");
        assert_eq!(safe_identifier("foo-bar"), "foo_bar");
        assert_eq!(safe_identifier(""), "_unknown");
    }
}
